### DISCUSSION TITLE
Adjust entity naming scheme to make it easier to make use of translated names

### DISCUSSION THREAD

**Discussion by emontnemery** [home-assistant/architecture#909] [Entity Models] [answered]
## Background

[Translations of entity components](https://developers.home-assistant.io/docs/internationalization/core/#name-of-entities) were recently added to core. In many cases entities could make use of those translations, this is in particular true for sensors and binary sensors.

Furthermore, the concept of [unnamed entities which share the name of its containing device](https://developers.home-assistant.io/docs/core/entity#has_entity_name-true-mandatory-for-new-integrations) was added some time ago. This is done by setting the entity's `name` attribute to `None` (the default).

## Problem
- Referencing translations for entity components is is fairly cumbersome; each integration needs to:
  - Set the `translation_key` property of entities
  - Add the translation key to its `strings.json` with a link to the corresponding translations in the entity component.

- While the concept of unnamed entities which share the name of its containing device is useful the way it's implemented is error prone and confusing.

## Suggestion
- Entity names default to a sentinel value indicating the entity is unnamed instead of to `None`
- Unnamed `binary_sensor`, `sensor` and possibly others with a device class should automatically fall back to the translations provided by the entity component
- Device name should be opt-in, by setting the entity's name attribute to `None`
  - During a deprecation period, if no device class and no name is set, warn, but keep current behavior (device name)

**Comment by balloob** [comment]
I think that specifically sensor and binary sensor benefit from being named after device class. I think that things like media players/garage doors, although there are device class names, generally the device itself has a name.

I think it's a good change overall but it should always do the same. So if name is omitted, HA will figure out a name based on device class or the current very bad default of `unnamed device` (for an entity). If a user wants to use the device name, explicitly pass `name=None` to entity description.

**Comment by frenck** [comment, marked as answer]
This is accepted and is currently being implemented.

../Frenck
